### PR TITLE
Tile plots (Heatmaps)

### PR DESCRIPTION
**Pull request by Kucharssim**
Fixes https://github.com/jasp-stats/jasp-issues/issues/1060

The heatmaps do not look that good with the palettes that are currently implemented in jaspGraphs, but that can be perhaps solved later.

I also have a question about column encoding, see the comment.

### CODE REVIEW

File: R/descriptives.R [hunk 1]
```diff
@@ -1541,3 +1555,92 @@ Descriptives <- function(jaspResults, dataset, options) {
...
 }
+
+
+.descriptivesHeatmaps <- function(container, dataset, variables, options) {
+  axesNames <- c(options[["heatmapHorizontal"]], options[["heatmapVertical"]])
+
+  # we are not ready to plot
+  if(length(variables) == "" || any(axesNames == "")) return()
+
+  axes <- .readDataSetToEnd(columns.as.factor = encodeColNames(axesNames))
```
Comment (Kucharssim) [inline R/descriptives.R:1566]: JASP was protesting when I did not do `encodeColNames()`, but I assume that the variable names should already come from qml encoded, or not? The values come from a `DropDown{}`.

**Review by vandenman — CHANGES_REQUESTED** [review]
A lot of functionality! I'm wondering a bit whether we should reorganize the interface, but perhaps that's something for later.

- [x] So the width and height of the plots is a bit off. I'm not sure wether
```r
plotSize <- c(100 + nLevels * 30) * c(options[["heatmapRectangleRatio"]], 1)
```
is a good idea here.

heatmap

I'd also set the x and y-axis breaks yourself, right now there are way too many ticks on the axes (which becomes noticeable if you copy the plot).



- [x] Something goes wrong with the state.
- add contcor1 and contcor2 as heatmap axes.
- add contNormal to variables
- a plot appears
- remove contNormal from variables
- the plot does **not** disappear

- [x] If you add the same variable for both heatmap axes, then any other variable gives an error:

image


- [x] I think a legend might be nice here, or at least an option for one. Because the color scale can represent intensity right?

File: R/descriptives.R (hunk 1 above)
Comment (vandenman) [reply R/descriptives.R:1566]: @JorisGoosen the column names should be encoded, even if they come from a `DropDown`, right? But if @Kucharssim had to resort to this, I'm not sure if they are.

**Comment by Kucharssim** [comment]
Thanks!

> So the width and height of the plots is a bit off. I'm not sure wether
> `plotSize <- c(100 + nLevels * 30) * c(options[["heatmapRectangleRatio"]], 1)`
>is a good idea here.

> If you add the same variable for both heatmap axes, then any other variable gives an error:

Ok so these things are maybe more of a problem of the ui than the analysis, I guess ideally the `DropDown` in this case would let you select only nominal/ordinal variables that were not already selected as `variables`, `splitby`, and the other axis. Any idea what how to do it? Otherwise I will just write checks for these cases.

**Comment by vandenman** [comment]
@boutinb any idea how to do what @Kucharssim describes? Ideally, we'd write something like:

```js
DropDown
{
	name: "heatmapHorizontal"
    // stuff
	source: [{ name: "allVariablesList", discard: ["variables", "splitby"], allowedColumns: ["ordinal", "nominal"]}]
}
```
We did something similar for JAGS once, but I'm not sure how general this syntax is.

File: R/descriptives.R (hunk 1 above)
Comment (JorisGoosen) [reply R/descriptives.R:1566]: They *should be* but it is possible there wasn't a proper meta added yet.
I'll make an issue for it.

**Comment by vandenman** [comment]
@Kucharssim sorry I forgot about this PR. Is this ready for another review?

**Comment by Kucharssim** [comment]
I forgot too, I will leave it for the next release I think, there are more pressing issues atm.

**Comment by AlexanderLyNL** [comment]
@Kucharssim could you please resolve the conflicts?

**Comment by Kucharssim** [comment]
Thanks to @boutinb the restrictions in the UI should now be working. This should cover the cases you did not like, @vandenman.

**Review by vandenman — CHANGES_REQUESTED** [review]
The code looks fine, just one minor comment.

However, when I try to open this PR as a dynamic module JASP (https://github.com/jasp-stats/jasp-desktop/commit/69fa56228161400300f123abd1988ea044c03184) segfaults at this line https://github.com/jasp-stats/jasp-desktop/blob/69fa56228161400300f123abd1988ea044c03184/Desktop/widgets/comboboxbase.cpp#L147

@boutinb any idea what is going on?

**Comment by Kucharssim** [comment]
> However, when I try to open this PR as a dynamic module JASP (jasp-stats/jasp-desktop@69fa562) segfaults at this line

If it's of any help, this does not occur to me on Mac with JASP build in Qt at this commit: https://github.com/jasp-stats/jasp-desktop/commit/9180752f01fe968927b35482d155eaf34df5e0c6 (I stick to this version because installing development modules using later versions do not work at all for me).

File: inst/qml/Descriptives.qml [hunk 2]
```diff
@@ -157,6 +157,51 @@ Form
...
+					addEmptyValue: true
+					source: [{ name: "allVariablesList", discard: ["variables", "splitby", "heatmapVertical"], use: "type=ordinal|nominal|nominalText"}]
+				}
+				DropDown
+				{
+					name: "heatmapVertical"
+					label: qsTr("Vertical axis:")
+					showVariableTypeIcon: true
+					addEmptyValue: true
+					source: [{ name: "allVariablesList", discard: ["variables", "splitby", "heatmapHorizontal"], use: "type=ordinal|nominal|nominalText"}]
```
Comment (boutinb) [inline inst/qml/Descriptives.qml:178]: You discard here heatmapHorizontal, but in heatmapVertical you discard heatmapHorizontal. There is a loop here that explains the crash experienced by @vandenman. I would also use a VariablesForm for this. 
I think the layout of the Plot section can be a bit polished

**Review by boutinb — CHANGES_REQUESTED** [review]

File: inst/qml/Descriptives.qml (hunk 2 above)
Comment (boutinb) [reply inst/qml/Descriptives.qml:178]: I would suggest to change the Plot Section in this way:
```
	Section
	{
		title: qsTr("Plots")
		columns: 1
		Group
		{
			title: "<b>" + qsTr("Basic plots") + "</b>"
			columns: 2
			CheckBox {				name: "plotVariables";			label: qsTr("Distribution plots");	id:	plotVariables					}
			CheckBox {				name: "plotCorrelationMatrix";	label: qsTr("Correlation plots");	id:	plotCorrelationMatrix			}

			Group
			{
				Layout.columnSpan: 2
				enabled: plotVariables.checked || plotCorrelationMatrix.checked

				indent:		true
				CheckBox {			name: "distPlotDensity";	label: qsTr("Display density")						}
				CheckBox {			name: "distPlotRug";		label: qsTr("Display rug marks")					}
				DropDown {
					name: "binWidthType"
					label: qsTr("Bin width type")
					indexDefaultValue: 0
					values:
						[
						{label: qsTr("Sturges"),				value: "sturges"},
						{label: qsTr("Scott"),					value: "scott"},
						{label: qsTr("Doane"),					value: "doane"},
						{label: qsTr("Freedman-Diaconis"),		value: "fd"	},
						{label: qsTr("Manual"),					value: "manual"	}
					]
					id: binWidthType
				}
				DoubleField
				{
					name:			"numberOfBins"
					label:			qsTr("Number of bins")
					defaultValue:	30
					min:			3;
					max:			10000;
					enabled:		binWidthType.currentValue === "manual"
				}
			}

			CheckBox {				name: "descriptivesIntervalPlot";label: qsTr("Interval plots")					}
			CheckBox {				name: "descriptivesQQPlot";		label: qsTr("Q-Q plots")						}
			CheckBox {				name: "descriptivesPiechart";	label: qsTr("Pie charts")						}
			CheckBox {				name: "descriptivesDotPlot";	label: qsTr("Dot plots")						}
		}


		Group
		{
			title: "<b>" + qsTr("Customizable plots") + "<b>"
			columns: 1
			DropDown
			{
				name: "colorPalette"
				label: qsTr("Color palette")
				indexDefaultValue: 0
				values:
				[
					{ label: qsTr("Colorblind"),		value: "colorblind"		},
					{ label: qsTr("Colorblind Alt."),	value: "colorblind3"	},
					{ label: qsTr("Viridis"),			value: "viridis"		},
					{ label: qsTr("ggplot2"),			value: "ggplot2"		},
					{ label: qsTr("Gray"),				value: "gray"			}
				]
			}

			CheckBox
			{
				name: "splitPlots";
				label: qsTr("Boxplots")
				Group {
					columns: 2
					Group {
						CheckBox {	name: "splitPlotBoxplot";		label: qsTr("Boxplot element"); checked: true	}
						CheckBox {	name: "splitPlotViolin";		label: qsTr("Violin element")					}
						CheckBox {	name: "splitPlotJitter";		label: qsTr("Jitter element")					}
					}
					Group {
						CheckBox {  name: "splitPlotColour";		label: qsTr("Use color palette")				}
						CheckBox {	name: "splitPlotOutlierLabel";	label: qsTr("Label outliers")					}
					}
				}
			}


			CheckBox
			{
				name: "scatterPlot";	label: qsTr("Scatter Plots")
				columns: 2
				RadioButtonGroup
				{
					name:	"graphTypeAbove";
					title:	qsTr("Graph above scatter plot")
					RadioButton { value: "density";		label: qsTr("Density");		checked: true	}
					RadioButton { value: "histogram";	label: qsTr("Histogram")					}
					RadioButton { value: "none";		label: qsTr("None")							}
				}
				RadioButtonGroup
				{
					name:	"graphTypeRight";
					title:	qsTr("Graph right of scatter plot")
					RadioButton { value: "density";		label: qsTr("Density");		checked: true	}
					RadioButton { value: "histogram";	label: qsTr("Histogram")					}
					RadioButton { value: "none";		label: qsTr("None")							}
				}
				CheckBox
				{
					name: "addSmooth"
					label: qsTr("Add regression line")
					checked: true
					RadioButtonGroup
					{
						name:	"regressionType";
						RadioButton { value: "smooth";	label: qsTr("Smooth");	checked: true	}
						RadioButton { value: "linear";	label: qsTr("Linear")					}
					}

					CheckBox
					{
						name: "addSmoothCI"
						label: qsTr("Show confidence interval")
						checked: true
						childrenOnSameRow: true
						CIField {	name: "addSmoothCIValue" }
					}
				}
				CheckBox
				{
					enabled: splitBy.count > 0
					name: "showLegend"
					label: qsTr("Show legend")
					checked: true
				}
			}
		}

		Group
		{
			title: "<b>" + qsTr("Tile Heatmap") + "<b>"

			VariablesForm
			{
				preferredHeight: 100 * preferencesModel.uiScale
				AvailableVariablesList
				{
					name: "heatMapVariables"
					source: [{ name: "allVariablesList", discard: ["variables", "splitby"], use: "type=ordinal|nominal|nominalText"}]
				}
				AssignedVariablesList
				{
					name: "heatmapHorizontal"
					label: qsTr("Horizontal axis:")
					singleVariable: true
				}
				AssignedVariablesList
				{
					name: "heatmapVertical"
					label: qsTr("Vertical axis:")
					singleVariable: true
				}
			}

			Group
			{
				columns: 2
				CheckBox
				{
					Layout.preferredWidth: 300
					name: "heatmapPlotValue"; label: qsTr("Display value"); childrenOnSameRow: false;
					DoubleField { name: "heatmapPlotValueSize"; label: qsTr("Relative text size"); negativeValues: false; defaultValue: 1 }
				}
				CheckBox { name: "heatmapLegend"; label: qsTr("Display legend")	}
				DoubleField { name: "heatmapRectangleRatio"; label: qsTr("Width to height ratio of tiles"); negativeValues: false; defaultValue: 1; Layout.columnSpan: 2}

				RadioButtonGroup
				{
					name: "heatmapStatisticContinuous"
					title: qsTr("Statistic to plot (scale variables)")
					RadioButton { value: "mean";		label: qsTr("Mean");	checked: true }
					RadioButton { value: "median";		label: qsTr("Median") }
					RadioButton { value: "identity";	label: qsTr("Value itself") }
					RadioButton { value: "length";		label: qsTr("Number of observations") }
				}
				RadioButtonGroup
				{
					name: "heatmapStatisticDiscrete"
					title: qsTr("Statistic to plot (nominal and ordinal variables)")
					RadioButton { value: "mode";		label: qsTr("Mode");	checked: true }
					RadioButton { value: "identity";	label: qsTr("Value itself") }
					RadioButton { value: "length";		label: qsTr("Number of observations") }
				}
			}
		}

	}
```

Comment (boutinb) [reply inst/qml/Descriptives.qml:178]: Then you get this:
Screenshot 2022-01-12 at 14 17 50

Comment (Kucharssim) [reply inst/qml/Descriptives.qml]: Thanks! Yeah that improved the interface as well, so

> I'm wondering a bit whether we should reorganize the interface, but perhaps that's something for later.

Is kind of tacked as well.

**Review by boutinb — APPROVED** [review]

**Comment by vandenman** [comment]
I like the new layout! However, this relation is a bit unintuitive. Perhaps the tile plots deserve their own color palette? Also, the default color palette (colorblind) is not really a good default when the variable is continuous. However, when the variable is nominal or ordinal, it's probably a good choice. 

image

I'm not sure what the best approach is here, but for now, this looks fine!

**Review by vandenman — APPROVED** [review]

**Pull request merged**